Added support for remote project in the LXD connection plugin

IshwarKanse opened the pull request:
… plugin

##### SUMMARY
For https://github.com/ansible-collections/community.general/issues/1441
Add support for specifying remote and project in the LXD connection plugin.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
community.general.lxd

##### ADDITIONAL INFORMATION
Tested the PR as below:

~~~
root@lxd:~# lxc --version
4.0.4

root@lxd:~# lxc remote list
+-----------------+------------------------------------------+---------------+-------------+--------+--------+
|      NAME       |                   URL                    |   PROTOCOL    |  AUTH TYPE  | PUBLIC | STATIC |
+-----------------+------------------------------------------+---------------+-------------+--------+--------+
| images          | https://images.linuxcontainers.org       | simplestreams | none        | YES    | NO     |
+-----------------+------------------------------------------+---------------+-------------+--------+--------+
| local (default) | unix://                                  | lxd           | file access | NO     | YES    |
+-----------------+------------------------------------------+---------------+-------------+--------+--------+
| myhost          | https://172.16.33.55:8443                | lxd           | tls         | NO     | NO     |
+-----------------+------------------------------------------+---------------+-------------+--------+--------+
| ubuntu          | https://cloud-images.ubuntu.com/releases | simplestreams | none        | YES    | YES    |
+-----------------+------------------------------------------+---------------+-------------+--------+--------+
| ubuntu-daily    | https://cloud-images.ubuntu.com/daily    | simplestreams | none        | YES    | YES    |
+-----------------+------------------------------------------+---------------+-------------+--------+--------+


root@lxd:~# lxc list myhost: --project=testproject
+-------+---------+------+------+-----------+-----------+
| NAME  |  STATE  | IPV4 | IPV6 |   TYPE    | SNAPSHOTS |
+-------+---------+------+------+-----------+-----------+
| test2 | RUNNING |      |      | CONTAINER | 0         |
+-------+---------+------+------+-----------+-----------+

root@lxd:~# cat inventory 
test2

[all:vars]
ansible_connection = community.general.lxd
ansible_lxd_remote = myhost
ansible_lxd_project = testproject

root@lxd:~# cat main.yml 
- hosts: all
  gather_facts: false

  collections:
  - community.general

  tasks:
    - name: Check the container hostname
      shell: hostname


root@lxd:~# ansible-playbook -i inventory main.yml -vvvv
ansible-playbook 2.10.4
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.8/dist-packages/ansible
  executable location = /usr/local/bin/ansible-playbook
  python version = 3.8.5 (default, Jul 28 2020, 12:59:40) [GCC 9.3.0]
No config file found; using defaults
setting up inventory plugins
host_list declined parsing /root/inventory as it did not pass its verify_file() method
script declined parsing /root/inventory as it did not pass its verify_file() method
auto declined parsing /root/inventory as it did not pass its verify_file() method
Parsed /root/inventory inventory source with ini plugin
Loading collection community.general from /root/.ansible/collections/ansible_collections/community/general
Loading callback plugin default of type stdout, v2.0 from /usr/local/lib/python3.8/dist-packages/ansible/plugins/callback/default.py
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: main.yml ****************************************************************************************************
Positional arguments: main.yml
verbosity: 4
connection: smart
timeout: 10
become_method: sudo
tags: ('all',)
inventory: ('/root/inventory',)
forks: 5
1 plays in main.yml

PLAY [all] ************************************************************************************************************
META: ran handlers

TASK [Check the container hostname] ********************************************************************************************
task path: /root/main.yml:8
<test2> ESTABLISH LXD CONNECTION FOR USER: root
<test2> EXEC /bin/sh -c 'echo ~root && sleep 0'
<test2> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1608443005.4742098-25659-4163528023467 `" && echo ansible-tmp-1608443005.4742098-25659-4163528023467="` echo /root/.ansible/tmp/ansible-tmp-1608443005.4742098-25659-4163528023467 `" ) && sleep 0'
<test2> Attempting python interpreter discovery
<test2> EXEC /bin/sh -c 'echo PLATFORM; uname; echo FOUND; command -v '"'"'/usr/bin/python'"'"'; command -v '"'"'python3.7'"'"'; command -v '"'"'python3.6'"'"'; command -v '"'"'python3.5'"'"'; command -v '"'"'python2.7'"'"'; command -v '"'"'python2.6'"'"'; command -v '"'"'/usr/libexec/platform-python'"'"'; command -v '"'"'/usr/bin/python3'"'"'; command -v '"'"'python'"'"'; echo ENDFOUND && sleep 0'
<test2> EXEC /bin/sh -c '/usr/bin/python3.5 && sleep 0'
Using module file /usr/local/lib/python3.8/dist-packages/ansible/modules/command.py
<test2> PUT /root/.ansible/tmp/ansible-local-25655pd7tvfk1/tmp0d8zlxqz TO /root/.ansible/tmp/ansible-tmp-1608443005.4742098-25659-4163528023467/AnsiballZ_command.py
<test2> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1608443005.4742098-25659-4163528023467/ /root/.ansible/tmp/ansible-tmp-1608443005.4742098-25659-4163528023467/AnsiballZ_command.py && sleep 0'
<test2> EXEC /bin/sh -c '/usr/bin/python3 /root/.ansible/tmp/ansible-tmp-1608443005.4742098-25659-4163528023467/AnsiballZ_command.py && sleep 0'
<test2> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1608443005.4742098-25659-4163528023467/ > /dev/null 2>&1 && sleep 0'
changed: [test2] => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python3"
    },
    "changed": true,
    "cmd": "hostname",
    "delta": "0:00:00.002845",
    "end": "2020-12-20 05:43:28.811535",
    "invocation": {
        "module_args": {
            "_raw_params": "hostname",
            "_uses_shell": true,
            "argv": null,
            "chdir": null,
            "creates": null,
            "executable": null,
            "removes": null,
            "stdin": null,
            "stdin_add_newline": true,
            "strip_empty_ends": true,
            "warn": true
        }
    },
    "rc": 0,
    "start": "2020-12-20 05:43:28.808690",
    "stderr": "",
    "stderr_lines": [],
    "stdout": "test2",
    "stdout_lines": [
        "test2"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP ************************************************************************************************************
test2                      : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
~~~

felixfontein commented on plugins/connection/lxd.py:
```diff
@@ -26,6 +26,18 @@
+            - Name of the LXD remote to use
+        default: local
+        vars:
+            - name: ansible_lxd_remote
```
```suggestion
            - name: ansible_lxd_remote
        version_added: 2.0.0
```

felixfontein commented on plugins/connection/lxd.py:
```diff
@@ -26,6 +26,18 @@
             - name: ansible_lxd_executable
+      remote:
+        description:
+            - Name of the LXD remote to use
```
```suggestion
            - Name of the LXD remote to use.
```

felixfontein commented on plugins/connection/lxd.py:
```diff
@@ -26,6 +26,18 @@
+            - name: ansible_lxd_remote
+      project:
+        description:
+            - Name of the LXD project to use
```
```suggestion
            - Name of the LXD project to use.
```

felixfontein commented on plugins/connection/lxd.py:
```diff
@@ -26,6 +26,18 @@
+            - Name of the LXD project to use
+        default: default
+        vars:
+            - name: ansible_lxd_project
```
```suggestion
            - name: ansible_lxd_project
        version_added: 2.0.0
```

felixfontein commented on plugins/connection/lxd.py:
```diff
@@ -70,7 +82,13 @@ def exec_command(self, cmd, in_data=None, sudoable=True):
-        local_cmd = [self._lxc_cmd, "exec", self._host, "--", self._play_context.executable, "-c", cmd]
+        local_cmd = [
+            self._lxc_cmd,
+            "--project", self.get_option("project"),
```
Does `--project default` always behaves the same as not being specified, for all versions of `lxc` supported by this plugin?

Akasurde replied:
Yes.

felixfontein commented on plugins/connection/lxd.py:
```diff
@@ -70,7 +82,13 @@ def exec_command(self, cmd, in_data=None, sudoable=True):
+            self._lxc_cmd,
+            "--project", self.get_option("project"),
+            "exec",
+            "%s:%s" % (self.get_option("remote"), self._host),
```
Same for `remote`: does `local:<host>` behave the same as `<host>` for all versions of `lxc` supported by this plugin?

stgraber replied:
Yeah, `local` has been a fixed remote in the LXD CLI ever since we've had the concept of remotes, so this shouldn't be a problem.

`--project` may be more problematic as LXD CLI prior to 3.6 will not know was a project is, similarly, a recent CLI interacting with an older LXD server will fail if that remote server is on a release prior to 3.6.

For this reason, it may be best to only pass `--project=` if the user specified a non-default project.

felixfontein reviewed (commented):
You need a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

IshwarKanse commented:
@Akasurde @stgraber @felixfontein I've removed the default for project option and added check to see if project is defined. Let me know if its good or do we need any other changes.

felixfontein commented on plugins/connection/lxd.py:
```diff
@@ -70,7 +83,21 @@ def exec_command(self, cmd, in_data=None, sudoable=True):
+                "exec",
+                "%s:%s" % (self.get_option("remote"), self._host),
+                "--",
+                self._play_context.executable, "-c", cmd]
```
This (and the other steps) could be greatly simplified (both for readability and maintanence in the future) by composing the list step for step, so that the `if` is only for adding the `--project` option.

I.e. something like
```.py
local_cmd = [self._lxc_cmd]
if self.get_option("project"):
    local_cmd.extend(["--project", self.get_option("project")])
local_cmd.extend([
    "exec",
    "%s:%s" % (self.get_option("remote"), self._host),
    "--",
    self._play_context.executable, "-c", cmd
])
```

IshwarKanse replied:
Thanks for the suggestion @felixfontein , I've made those changes. :)

felixfontein commented:
@stgraber @Akasurde what do you think?

stgraber commented:
Looks good to me!

Akasurde commented:
recheck

felixfontein commented:
@IshwarKanse thanks for implementing this!
@stgraber @Akasurde thanks for reviewing!
